src/standalone/Snow/neural_snow/ModelTools.jl

```diff
@@ -0,0 +1,322 @@
+- `timeseries::DataFrame`: The input data frame used to generate predictions, including a time variable.
+- `dt::Period`: The unit timestep present in the dataframe (i.e. daily dataframe, dt = Day(1) or Second(86400)).
+- `predictvar::Symbol`: The variable to predict from the timeseries. Default is :z.
+- `timevar::Symbol`: The variable giving the time of each forcing. Default is :date.
+- `inputvars::Vector{Symbol}`: The variables (in order), to extract from the data to use for predictions.
+Default is [:z, :SWE, :rel_hum_avg, :sol_rad_avg, :wind_speed_avg, :air_temp_avg, :dprecipdt_snow] like the paper.
+- `dtype::Type`: The data type required for input to the model. Default is Float32.
+- `hole_thresh::Int`: The acceptable number of "holes" in the timeseries for the model to skip over. Default is 30.
+"""
+function make_timeseries(
+    model,
+    timeseries::DataFrame,
+    dt::Period;
+    predictvar::Symbol = :z,
+    timevar::Symbol = :date,
+    inputvars::Vector{Symbol} = [
+        :z,
+        :SWE,
+        :rel_hum_avg,
+        :sol_rad_avg,
+        :wind_speed_avg,
+        :air_temp_avg,
+        :dprecipdt_snow,
+    ],
+    dtype::Type = Float32,
+    hole_thresh::Int = 30,
+)
+    forcings = Matrix{dtype}(select(timeseries, inputvars))
+    pred_idx = findfirst(inputvars .== predictvar)
+    check_dates =
+        (timeseries[2:end, timevar] - timeseries[1:(end - 1), timevar]) ./ dt
+    pred_vals = zeros(nrow(timeseries) - 1)
+    pred_series = zeros(nrow(timeseries))
+    pred_series[1] = timeseries[1, predictvar]
+    countresets = 0
+    for j in 2:length(pred_series)
+        input = forcings[j - 1, :]
+        input[pred_idx] = pred_series[j - 1]
+        pred = evaluate(model, input)[1]
+        pred_vals[j - 1] = pred
+        nperiods = check_dates[j - 1]
+        new_val = pred_series[j - 1] + nperiods * Dates.value(Second(dt)) * pred
+        pred_series[j] =
+            (nperiods <= hole_thresh) ? max(0.0, new_val) :
+            timeseries[j, predictvar]  #the "max" is only in the case of holes for a non-negative system and does not generalize
+        #pred_series[j] = (nperiods <= hole_thresh) ? new_val : true_series[j]  #for showing no thresholds
+        if nperiods > hole_thresh
+            countresets += 1
+        end
+    end
+    return pred_series, pred_vals, countresets
+end
+
+
+"""
+    custom_loss(x, y, model, n1, n2)
+
+Creates a loss function to be used during training specified by two hyperparameters n1, n2, as outlined in the paper.
+
+# Arguments
+- `x`: the input values.
+- `y`: output values to compare to.
+- `model`: the model over which to evaluate the loss function.
+- `n1::Int`: the hyperparameter dictating the scaling of mismatch error.
+- `n2::Int`: the hyperparameter dictating the weighting of a given mismatch error by the target magnitude.
+"""
+function custom_loss(x, y, model, n1, n2)
+    sum(abs.(model(x) - y) .^ n1 .* (1.0 .+ abs.(y) .^ n2)) ./ length(y)
```

Comment:
you might check, but I think 1.0 +abs(x) will return Float64 even if x is Float32. Swapping to 1 + abs(x) will preserve the type